modules/auxiliary/cloud/aws/enum_ssm.rb

```diff
@@ -0,0 +1,153 @@
+                         ),
+        'Author'      => [
+          'Aaron Soto <aaron.soto@rapid7.com>', # EC2 enum module
+          'RageLtMan <rageltman[at]sempervictus>' # SSM stuff
+        ],
+        'License'     => MSF_LICENSE
+      )
+    )
+
+    register_options(
+      [
+        OptBool.new('GET_SSM_SESSION', [true, 'Automatically get SSM sessions once found', false]),
+        OptInt.new('LIMIT', [false, 'Only return the specified number of results from each region']),
+        OptString.new('FILTER_EC2_ID', [false, 'Look for specific EC2 instance ID']),
+        OptString.new('REGION', [false, 'AWS Region (eg. "us-west-2")']),
+        OptString.new('ACCESS_KEY_ID', [true, 'AWS Access Key ID (eg. "AKIAXXXXXXXXXXXXXXXX")', '']),
+        OptString.new('SECRET_ACCESS_KEY', [true, 'AWS Secret Access Key (eg. "CA1+XXXXXXXXXXXXXXXXXXXXXX6aYDHHCBuLuV79")', ''])
+      ]
+    )
+  end
+
+  def handle_aws_errors(e)
+    if e.class.module_parents.include?(Aws)
+      fail_with(Failure::UnexpectedReply, e.message)
+    else
+      raise e
+    end
+  end
+
+  def enumerate_regions
+    regions = []
+
+    ec2 = Aws::EC2::Resource.new(
+      region: 'us-east-1',
+      access_key_id: datastore['ACCESS_KEY_ID'],
+      secret_access_key: datastore['SECRET_ACCESS_KEY']
+    )
+
+    ec2_regions = ec2.client.describe_regions.data.regions
+    ec2_regions.each do |r|
+      regions.append(r.region_name)
+    end
+
+    regions
+  end
+
+  def run
+    regions = datastore['REGION'] ? [datastore['REGION']] : regions = enumerate_regions()
+    credentials = ::Aws::Credentials.new(datastore['ACCESS_KEY_ID'], datastore['SECRET_ACCESS_KEY'])
+    regions.each do |region|
+      begin
+        vprint_status "Checking #{region}..."
+        client = ::Aws::SSM::Client.new(
+          region: region,
+          credentials: credentials,
+        )
+        inv_params = { filters: [
+          {
+            key: "AWS:InstanceInformation.InstanceStatus",
+            values: ["Terminated"],
+            type: "NotEqual",
+          },
+          {
+            key: "AWS:InstanceInformation.ResourceType",
+            values: ['EC2Instance'],
+            type: "Equal",
+          }
+        ]}
+
+        inv_params[:filters] << {
+          key: "AWS:InstanceInformation.InstanceId",
+          values: [datastore['FILTER_EC2_ID']],
+          type: "Equal",
+        } if datastore['FILTER_EC2_ID']
+
+        ssm_ec2 = client.get_inventory(inv_params).entities.map {|e| e.data["AWS:InstanceInformation"].content}.flatten
+        ssm_ec2 = ssm_ec2[0...datastore['LIMIT']] if datastore['LIMIT']
+        ssm_ec2.each do |ssm_host| 
+          vprint_good JSON.pretty_generate(ssm_host)
+          if datastore['GET_SSM_SESSION']
+            socket = get_ssm_socket(client, ssm_host['InstanceId'])
+            start_session(self, "AWS SSM #{datastore['ACCESS_KEY_ID']} (#{ssm_host['InstanceId']})", datastore, false, socket.lsock)
+          end
```

Comment:
Somewhere in here I appear to be getting another error.

```
msf6 auxiliary(cloud/aws/enum_ssm) > run

[*] Checking us-east-2...
[+] {
  "AgentType": "amazon-ssm-agent",
  "AgentVersion": "3.1.1732.0",
  "ComputerName": "ip-172-#-#-#.us-east-2.compute.internal",
  "InstanceId": "i-##########",
  "InstanceStatus": "Active",
  "IpAddress": "172.#.#.#",
  "PlatformName": "Amazon Linux",
  "PlatformType": "Linux",
  "PlatformVersion": "2",
  "ResourceType": "EC2Instance"
}
[-] Auxiliary aborted due to failure: unexpected-reply: Document with name SSM-SessionManagerRunShell does not exist.
[*] Auxiliary module execution completed
msf6 auxiliary(cloud/aws/enum_ssm) > 
```

Reply:
Likely dont have permissions per https://docs.aws.amazon.com/systems-manager/latest/userguide/getting-started-sessiondocumentaccesscheck.html

Reply:
@smcintyre-r7 - how do you want to slice this? Should we auto-create the relevant docs or have an option to do so?

Reply:
I ended up following [this guide](https://docs.aws.amazon.com/systems-manager/latest/userguide/getting-started-create-preferences-cli.html) to create the doc. That could go in the module docs but isn't really related to how all the code is functioning so I'm not terribly concerned about it. Ideally we'd document it with payloads but payloads don't have module docs like other modules do. I can write up some of my setup notes in this PR so they're not lost until we figure out where to put them long term.